extending.md

```diff
@@ -0,0 +1,132 @@
+# Extending `smart_open`
+
+This document targets potential contributors to `smart_open`.
```

Comment:
What's about actions in case if I want to add new format, but without pushing them to smart_open (for example, reader for proprietary stuff in the company, useless for open-source).